File: NBlast.Api/Controllers/IndexerController.fs
namespace NBlast.Api.Controllers

open NBlast.Api.Models
open NBlast.Api.Async
open NBlast.Storage
open NBlast.Storage.Core.Index
open System.Web.Http
open System.Net.Http.Formatting
open System.Web.Http.Cors

[<RoutePrefix("api/indexer")>]
[<EnableCors("*", "*", "GET")>]
type IndexerController(queueKeeper: IIndexingQueueKeeper) =
    inherit ApiController()

    [<HttpPost>]
    [<Route("index")>]
    member me.Index (model: LogModel) =
        if (me.ModelState.IsValid) then
            queueKeeper.Enqueue(model)
            me.Ok(model) :> IHttpActionResult
        else
            me.BadRequest(me.ModelState) :> IHttpActionResult
    
    [<HttpGet>]
    [<Route("queue-count")>]
    member me.QueueCount() =
        me.Ok(queueKeeper.Count())

    [<HttpGet>]
    [<Route("queue-content/{top}")>]
    member me.QueueContent(top: int) = queueKeeper.PeekTop(top)

File: NBlast.Rest/Controllers/Api/IndexerController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web.Http;
using System.Web.Http.Cors;
using System.Web.Http.Results;
using NBlast.Rest.Async;
using NBlast.Rest.Model.Write;
using Serilog;

namespace NBlast.Rest.Controllers.Api
{
    [EnableCors("*", "*", "GET")]
    [RoutePrefix("api/indexer")]
    public class IndexerController: ApiController
    {
        private readonly IIndexingQueueKeeper _indexingQueueKeeper;
        private static readonly ILogger Logger = Log.Logger.ForContext<IndexerController>();

        public IndexerController(IIndexingQueueKeeper indexingQueueKeeper)
        {
            if (indexingQueueKeeper == null) throw new ArgumentNullException(nameof(indexingQueueKeeper));
            _indexingQueueKeeper = indexingQueueKeeper;
        }


        [HttpPost]
        [Route("index")]
        public IHttpActionResult Index(LogEvent @event)
        {
            if (!ModelState.IsValid)
            {
                Logger.Debug($"Log model {@event} is INVALID");
                return BadRequest(ModelState);
            }
            _indexingQueueKeeper.Enqueue(@event); // TODO Maybe clone it before
            return Ok(@event);
        }

        [HttpPost]
        [Route("index-events")]
        public IHttpActionResult Index(LogEvents eventsKeeper)
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }
            eventsKeeper.Events.ToList().ForEach(e => _indexingQueueKeeper.Enqueue(e));
            return Ok();
        }

        [HttpGet]
        [Route("queue-count")]
        public int QueueCount() => _indexingQueueKeeper.Count();

        [HttpGet]
        [Route("queue-content/{top}")]
        public IReadOnlyList<LogEvent> QueueContent(int top) =>
            _indexingQueueKeeper.PeekTop(top);
    }
}